File: cpp/leetcode/173.cpp
class BSTIterator {
    stack<TreeNode *> stk;
public:
    BSTIterator(TreeNode *root) {
        pushAll(root);
    }

    bool hasNext() {
        return !stk.empty();
    }

    int next() {
        TreeNode *node = stk.top();
        stk.pop();
        pushAll(node->right);
        return node->val;
    }

private:
    void pushAll(TreeNode *node) {
        while(node){
            stk.push(node);
            node = node->left;
        }
    }
};

File: python/leetcode/173.py
class BSTIterator:
    def __init__(self, root):
        self.stack = []
        self.pushAll(root)

    def hasNext(self):
        return self.stack

    def next(self):
        node = self.stack.pop()
        self.pushAll(node.right)
        return node.val
        
    def pushAll(self, node):
        while node is not None:
            self.stack.append(node)
            node = node.left